Update proposal based on feedback (#5)
g-gaston/cluster-api: docs/proposals/20240807-in-place-updates.md

#### Clean separation of concern

It is the responsibility of the extension to decide if it can perform changes in-place and to perform these changes on a single machine. If the extension decides that it cannot perform changes in-place, CAPI will fall back to rollout.

The responsibility to determine which machine should be rolled out as well as the responsibility to handle rollout options like MaxSurge/MaxUnavailable will remain on the controllers owning the machine (e.g. KCP, MD controller). 

### Goals

- Enable the implementation of pluggable update extensions.
- Allow users to update Kubernetes clusters using pluggable External Update Extension.
- Maintain a coherent user experience for both rolling and in-place updates.
- Support External Update Extensions for both Control Plane (KCP or others) and MachineDeployment controlled machines.

### Non-Goals/Future work

- To provide rollbacks in case of an in-place update failure. Failed updates need to be fixed manually by the user on the machine or by replacing the machine.
- Introduce any changes to KCP (or any other control plane provider), MachineDeployment, MachineSet, Machine APIs.

> - Ammend the desired state to something that the registered updaters can cover or register additional updaters capable of handling the desired changes.

Reviewer: There was a comment about this sentence being hard to understand, maybe we can remove it?

Author: Moved it to `Non-Goals/Future work` section as per [suggestion](https://github.com/kubernetes-sigs/cluster-api/pull/11029/files#r2008927060). As per Danil's comment below, I'll indicate this is indeed a future goal. Let me know if you still think we should remove this though.

> - Allow in-place updates for single-node clusters without the requirement to reprovision hosts.

Reviewer 2: ```suggestion
- Allow in-place updates for single-node clusters without the requirement to reprovision hosts (future goal).
```
Since it will likely be covered anyways.

## Proposal

> We propose to extend upgrade workflows to call External Update Extensions, if defined.

Reviewer 2: I feel like the original definition is clear. This change implies mentioning that functionality will be behind FG. Maybe:

```suggestion
We propose a pluggable update strategy architecture that allows to register External Update Extension to optionally handle the update process.
```
I think it is sufficient to drop the change, as code organization is addressed below.

> The MachineDeployment controller updates machines in place in a very similar way to rolling updates: by creating a new MachineSet and moving the machines from the old MS to the new one. We want to stress that the Machine objects won't be deleted and recreated like in the current rolling strategy. The MachineDeployment will just update the OwnerRefs and labels, effectively moving the existing Machine object from one MS to another. The number of machines moved at once might be made configurable on the MachineDeployment in the same way `maxSurge` and `maxUnavailable` control this for rolling updates.

Reviewer 2: ```suggestion
```
Needs to be removed, will be established during implementation

The controller will trigger updaters by hitting a RuntimeHook endpoint (eg. `/UpdateMachine`). The updater could respond saying "update completed", "update failed" or "update in progress" with an optional "retry after X seconds". The CAPI controller will continuously poll the status of the update by hitting the same endpoint until it reaches a terminal state.

CAPI expects the `/UpdateMachine` endpoint of an updater to be idempotent: for the same Machine with the same spec, the endpoint can be called any number of times (before and after it completes), and the end result should be the same. CAPI guarantees that once an `/UpdateMachine` endpoint has been called once, it won't change the Machine spec until the update either completes or fails.

Once all of the updaters are complete, the Machine controller will mark machine as done. If the update fails, this will be reflected in the Machine status.

From this point on, the `KCP` or `MachineDeployment` controller will take over and set the `UpToDate` condition to `True`.

> Note: We might revisit which controller should set `UpToDate` during implementation, because we have to make sure there are no race conditions that can lead to reconcile failures, but apart from the ownership of this operation, the workflows described in this doc should not be impacted.

Reviewer 2: 👍🏻

Since the fallback to machine replacement is a default strategy and always enabled, the MachineDeployment controller proceeds with the rollout process as it does today, replacing the old machines with new ones.

### API Changes

> *All functionality related to In-Place Updates will be available only if the `InPlaceUpdates` feature flag is set to true.*

Reviewer 2: I think this part of the statement still needs to be present.
